platform/rct/hal/usb_stm32/USB-CDC_device_stm32.c

```diff
@@ -1,52 +1,250 @@
+	volatile uint32_t USB_Rx_ptr_in;
+	volatile uint32_t USB_Rx_ptr_out;
+	volatile int _init_flag;
+	xSemaphoreHandle _rlock;
+	xSemaphoreHandle _wlock;
+} usb_state;
+
+/* ST's USB hardware doesn't include an internal pullup resistor on
+ * USB_DP for some odd reason and instead relies on an externally
+ * gated (via a set of transistors) pull up attached to a
+ * USB_DISCONNECT pin. As this is only used during startup (we don't
+ * do any dynamic re-enumeration), we can shortcut this by having a
+ * static pullup, and simply causing a glitch on the line by forcing
+ * USB_DP low before we initialize the rest of the USB hardware. It's
+ * a little hacky, but doesn't seem to cause any ill effects. */
+static int USB_CDC_force_reenumeration(void)
+{
+	GPIO_InitTypeDef GPIO_InitStructure;
+
+	RCC_AHBPeriphClockCmd(RCC_AHBPeriph_GPIOA, ENABLE);
+	GPIO_InitStructure.GPIO_Pin = GPIO_Pin_12;
+	GPIO_InitStructure.GPIO_Speed = GPIO_Speed_50MHz;
+	GPIO_InitStructure.GPIO_Mode = GPIO_Mode_OUT;
+	GPIO_InitStructure.GPIO_OType = GPIO_OType_PP;
+	GPIO_InitStructure.GPIO_PuPd = GPIO_PuPd_NOPULL;
+	GPIO_Init(GPIOA, &GPIO_InitStructure);
+
+ 	GPIO_ResetBits(GPIOA, GPIO_Pin_12);
+
+	GPIO_InitStructure.GPIO_OType = GPIO_OType_OD;
+	GPIO_Init(GPIOA, &GPIO_InitStructure);
+
+}
 
-#include <usbd_cdc_core.h>
-#include <usbd_usr.h>
-#include <usbd_conf.h>
-#include <usbd_desc.h>
-#include <usbd_cdc_vcp.h>
+/* Public API */
+int USB_CDC_device_init(const int priority)
+{
+	/* Perform a full USB reset */
+	USB_CDC_force_reenumeration();
+
+	/* Create read/write mutexes */
+	usb_state._rlock = xSemaphoreCreateMutex();
+	usb_state._wlock = xSemaphoreCreateMutex();
+
+	if (usb_state._rlock == NULL || usb_state._wlock == NULL) {
+		return -1;
+	}
+
+	Set_System();
+	Set_USBClock();
+
+	USB_Interrupts_Config();
+
+	USB_Init();
 
-static volatile int _init_flag = 0;
-USB_OTG_CORE_HANDLE USB_OTG_dev __attribute__ ((aligned (4)));
+	usb_state._init_flag = 1;
 
-int USB_CDC_device_init()
+	return 0;
+}
+
+void USB_CDC_Write(portCHAR *cByte, int len)
 {
-    vcp_setup();
+	xSemaphoreTake(usb_state._wlock, portMAX_DELAY);
+	while (len--) {
+		/* If the transmit buffer is full, wait until the USB
+		 * hardware has had time to flush it out */
+		uint16_t usb_buf_elements = USB_BUF_ELTS(usb_state.USB_Tx_ptr_in,
+							 usb_state.USB_Tx_ptr_out,
+							 VIRTUAL_COM_PORT_DATA_SIZE);
+
+		while (usb_buf_elements == VIRTUAL_COM_PORT_DATA_SIZE - 1) {
+			portYIELD();
+			usb_buf_elements = USB_BUF_ELTS(usb_state.USB_Tx_ptr_in,
+							usb_state.USB_Tx_ptr_out,
+							VIRTUAL_COM_PORT_DATA_SIZE);
+		}
 
-    /* Initialize the USB hardware */
-    USBD_Init(&USB_OTG_dev,
-              USB_OTG_FS_CORE_ID,
-              &USR_desc,
-              &USBD_CDC_cb,
-              &USR_cb);
-    _init_flag = 1;
+		usb_state.USB_Tx_Buffer[usb_state.USB_Tx_ptr_in] = *cByte++;
 
-    return 0;
+		/* Handle wrapping */
+		if(usb_state.USB_Tx_ptr_in == VIRTUAL_COM_PORT_DATA_SIZE - 1) {
```

Comment:
I think getting something working is the highest priority.  But #2 priority will be trimming away all unnecessary memory usage just because of the limited space we are working with.  I would suggest filing some issues against this so its not forgotten as we move forward.  Unfortunately in software nothing is more permanent that temporary.